src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java

```diff
@@ -0,0 +1,319 @@
+
+    private static final Pattern IDENTIFIER = Pattern.compile("[_a-zA-Z][_a-zA-Z0-9]+");
+
+    private final SQLConnection jdbcContext;
+
+    public EscapeSyntaxParser(SQLConnection jdbcContext) {
+        this.jdbcContext = jdbcContext;
+    }
+
+    /**
+     * Performs escape processing for SQL queries. It translates
+     * sql text with optional escape expressions such as {@code {fn abs(-1)}}.
+     *
+     * <p>
+     * Comments inside SQL text can be eliminated as parsing goes using preserveComments
+     * flag. Hence, Comments inside escape syntax are always omitted regardless of
+     * the flag, though.
+     *
+     * @param sql              SQL text to be processed
+     * @param preserveComments flag indicating should comments be kept
+     *
+     * @return native SQL query
+     *
+     * @throws SQLSyntaxErrorException if any syntax error happened
+     */
+    public String translate(String sql, boolean preserveComments) throws SQLSyntaxErrorException {
+        StringBuilder nativeSql = new StringBuilder(sql.length());
+        StringBuilder escapeBuffer = new StringBuilder();
+        StringBuilder activeBuffer = nativeSql;
+        LinkedList<Integer> escapeStartPositions = new LinkedList<>();
+
+        int i = 0;
+        while (i < sql.length()) {
+            char currentChar = sql.charAt(i);
+            switch (currentChar) {
+            case '\'':
+            case '"':
+                int endOfString = seekEndOfRegion(sql, i, "" + currentChar, "" + currentChar);
+                if (endOfString == -1) {
+                    throw new SQLSyntaxErrorException(
+                        "Not enclosed string literal or quoted identifier at position " + i,
+                        SQLStates.SYNTAX_ERROR.getSqlState()
+                    );
+                }
+                activeBuffer.append(sql, i, endOfString + 1);
+                i = endOfString + 1;
+                break;
+
+            case '/':
+            case '-':
+                int endOfComment;
+                if (currentChar == '/') {
+                    endOfComment = seekEndOfRegion(sql, i, "/*", "*/");
+                    if (endOfComment == -1) {
+                        throw new SQLSyntaxErrorException(
+                            "Open block comment at position " + i, SQLStates.SYNTAX_ERROR.getSqlState()
+                        );
+                    }
+                } else {
+                    endOfComment = seekEndOfRegion(sql, i, "--", "\n");
+                    if (endOfComment == -1) {
+                        endOfComment = sql.length() - 1;
+                    }
+                }
+                if (i == endOfComment) {
+                    activeBuffer.append(currentChar);
+                    i++;
+                } else {
+                    if (activeBuffer == nativeSql && preserveComments) {
+                        nativeSql.append(sql, i, endOfComment + 1);
+                    }
```

Comment:
removed

Reply:
```diff
@@ -101,9 +104,6 @@ public class EscapeSyntaxParser {
                     activeBuffer.append(currentChar);
                     i++;
                 } else {
-                    if (activeBuffer == nativeSql && preserveComments) {
-                        nativeSql.append(sql, i, endOfComment + 1);
-                    }
                     i = endOfComment + 1;
                 }
                 break;
```

So it seems that the problem with `token1/*comment*/token2` still here. We should add a whitespace or a newline if we'll strip comments.

Aside of that we discussed this a bit with @Korablev77 (Tarantool/SQL guy) and we both doubt that stripping of comments is the good idea if it is not required by the standard and is not requested by a user. The point here is to provide good debugging facilities: say, report correct line number and full text of a request in case of an error.

I propose to don't strip anything by default and maybe add an option in a future (if we'll see a demand).

BTW, why comments in escape syntax are stripped unconditionally? I would say that we should not strip them too as well (but I don't know if there any problems with it).

Reply:
Hm, AFAIS, now escape syntax comments are preserved unconditionally. So only the question remains is about comments outside of escape syntax.

Reply:
> Hm, AFAIS, now escape syntax comments are preserved unconditionally. So only the question remains is about comments outside of escape syntax.

My bad, I misinterpreted the updated code. Yep, now comments are removed everywhere. My proposal is to preserve them everywhere.

Reply:
Agree, but I have one objection here. It touches comments within escape expressions. 

- I found out there are cases when we cannot definitely keep comments as-is. Let's say, an expression like `select {fn database(/* get db name */)}`; here the transformed query will be `select 'universe'`, thus, it makes a correct transformation impossible, IMO.

- Next, preserving comments is working now till we just omitting `{}` braces (oj, limit, etc.); but we cannot guarantee that escape syntax won't use a more complex transformation in the future that won't be compatible with comments as in the previous point. 

- JDBC does not allow or deny comments inside `{}` explicitly. But the syntax is more or less strict (i.e. the scalar function is described as `{fn <function-name> (argument list)}`) and I think here a more correct way is to inform users that we follow the escape syntax in a stricter way here and the comments within *escapes* are not supported. it might cause a syntax exception or silent filtering out. I've chosen the second option: write down attention in Javadoc (+JDCB doc) that SQL comments won't be present after the processing because they can impact on the correct transformation.

To sum it up, it's ok to preserve the comments outside the escapes (debugging and other purposes) but there are a few (lot?) corner cases if we try to handle comments in a block being
processed by the driver. I'd like to always replace them with whitespaces in this case.

An example: `select /*may return guest*/ {fn user(/*current user*)}` becomes `select /*may return guest*/ 'guest'`.

Reply:
`/* <...> */` and `-- <...> EOL` comments cannot be misinterpreted as some other SQL clause and vice versa. The only case when we're unable to keep a comment as is is when we replace an entire expression that contains it on a driver side: this is the valid transformation. So I don't see how comments may impact correctness of a transformation.

JDBC is naturally based on SQL and, say, expression `{fn <function name> (<argument list>)` treats `<function name>` as identifier using SQL rules, tokenizes `(`, `)` and `<argument list>` as SQL requires, assumes that `<argument list>` is a list of valid comma separated SQL expressions. At least it means that `<argument list>` can contain a comment, but I think it is more natural to support comments for escape syntax in SQL way: on tokenizer level.

If we'll need to make more complex transformations, it would be better to work on extraction of SQL parser from tarantool (https://github.com/tarantool/tarantool/issues/3319) and transform an AST rather then a text on driver's side.

I have two possible reasons in the mind to keep comments:

* A user may have deep nested joins and may want to comment them.
* A kind of unification on tokenizer level: the same rules for SQL everywhere.

But we can look what other connectors doing. Maybe my way to think about those aspects is not popular.